File: gcp_cost_optimizer_agent/go/tools/compute.go
package tools

import (
	"fmt"
	"sort"
	"strings"

	compute "cloud.google.com/go/compute/apiv1"
	computepb "cloud.google.com/go/compute/apiv1/computepb"
	"google.golang.org/adk/tool"
	"google.golang.org/adk/tool/functiontool"
	"google.golang.org/api/iterator"
)

type ListRunningVMsArgs struct {
	ProjectID string `json:"project_id" description:"The GCP project ID (e.g. 'my-project')."`
}

type VMInfo struct {
	Name        string  `json:"name"`
	MachineType string  `json:"machine_type"`
	Status      string  `json:"status"`
	InternalIP  *string `json:"internal_ip"`
	ExternalIP  *string `json:"external_ip"`
}

type ZoneSummaryItem struct {
	Zone  string `json:"zone"`
	Count int    `json:"count"`
}

type ListRunningVMsResult struct {
	Total   int                 `json:"total"`
	ByZone  map[string][]VMInfo `json:"by_zone"`
	Summary []ZoneSummaryItem   `json:"summary"`
}

func NewListRunningVMsTool() (tool.Tool, error) {
	return functiontool.New(functiontool.Config{
		Name:        "list_running_vms",
		Description: "List running Compute Engine VM instances in a project with machine types. Use this to identify oversized or idle VMs.",
	}, ListRunningVMs)
}

func ListRunningVMs(ctx tool.Context, args ListRunningVMsArgs) (ListRunningVMsResult, error) {
	client, err := compute.NewInstancesRESTClient(ctx)
	if err != nil {
		return ListRunningVMsResult{}, fmt.Errorf("creating compute instances client: %w", err)
	}
	defer client.Close()

	filterStr := "status=RUNNING"
	req := &computepb.AggregatedListInstancesRequest{
		Project: args.ProjectID,
		Filter:  &filterStr,
	}

	byZone := make(map[string][]VMInfo)
	it := client.AggregatedList(ctx, req)
	for {
		pair, err := it.Next()
		if err == iterator.Done {
			break
		}
		if err != nil {
			return ListRunningVMsResult{}, fmt.Errorf("iterating aggregated instances: %w", err)
		}

		instancesScopedList := pair.Value
		if len(instancesScopedList.Instances) == 0 {
			continue
		}

		// zone is in the format "zones/us-central1-a"
		zoneName := pair.Key
		if strings.HasPrefix(zoneName, "zones/") {
			zoneName = strings.TrimPrefix(zoneName, "zones/")
		}

		var vms []VMInfo
		for _, vm := range instancesScopedList.Instances {
			vms = append(vms, formatInstance(vm))
		}
		byZone[zoneName] = vms
	}

	var summary []ZoneSummaryItem
	total := 0
	for z, vms := range byZone {
		summary = append(summary, ZoneSummaryItem{
			Zone:  z,
			Count: len(vms),
		})
		total += len(vms)
	}

	sort.Slice(summary, func(i, j int) bool {
		return summary[i].Count > summary[j].Count
	})

	return ListRunningVMsResult{
		Total:   total,
		ByZone:  byZone,
		Summary: summary,
	}, nil
}

func formatInstance(vm *computepb.Instance) VMInfo {
	var internalIP, externalIP *string

	for _, iface := range vm.NetworkInterfaces {
		internalIP = iface.NetworkIP
		for _, accessConfig := range iface.AccessConfigs {
			if accessConfig.NatIP != nil && *accessConfig.NatIP != "" {
				externalIP = accessConfig.NatIP
			}
		}
		break // First interface only as in python code
	}

	machineType := ""
	if vm.MachineType != nil {
		parts := strings.Split(*vm.MachineType, "/")
		machineType = parts[len(parts)-1]
	}

	status := ""
	if vm.Status != nil {
		status = *vm.Status
	}

	name := ""
	if vm.Name != nil {
		name = *vm.Name
	}

	return VMInfo{
		Name:        name,
		MachineType: machineType,
		Status:      status,
		InternalIP:  internalIP,
		ExternalIP:  externalIP,
	}
}

File: gcp_cost_optimizer_agent/python/tools/compute.py
"""Compute Engine tool — list running VM instances in a project."""

from __future__ import annotations

from google.cloud import compute_v1


def list_running_vms(project_id: str) -> dict:
    """List all running VM instances in a project.

    Args:
        project_id: The GCP project ID (e.g. "my-project").

    Returns:
        A dict with:
          - "total": number of running instances
          - "by_zone": dict mapping zone -> list of VM dicts, each with:
              name, machine_type, status, internal_ip, external_ip (if any)
          - "summary": list of {"zone": str, "count": int} sorted by count desc
    """
    client = compute_v1.InstancesClient()
    request = compute_v1.AggregatedListInstancesRequest(
        project=project_id,
        filter="status=RUNNING",
    )

    by_zone: dict[str, list[dict]] = {}
    
    for zone, instances_scoped_list in client.aggregated_list(request=request):
        vms = instances_scoped_list.instances
        if not vms:
            continue
        zone_name = zone.removeprefix("zones/")
        by_zone[zone_name] = [_format_instance(vm) for vm in vms]

    summary = sorted(
        [{"zone": z, "count": len(vms)} for z, vms in by_zone.items()],
        key=lambda x: x["count"],
        reverse=True,
    )

    return {
        "total": sum(len(v) for v in by_zone.values()),
        "by_zone": by_zone,
        "summary": summary,
    }


def _format_instance(vm: compute_v1.Instance) -> dict:
    """Extract key fields from a Compute Instance into a plain dict."""
    internal_ip = None
    external_ip = None
    for iface in vm.network_interfaces:
        internal_ip = iface.network_i_p
        for config in iface.access_configs:
            if config.nat_i_p:
                external_ip = config.nat_i_p
        break  # first interface is enough

    return {
        "name": vm.name,
        "machine_type": vm.machine_type.split("/")[-1],
        "status": vm.status,
        "internal_ip": internal_ip,
        "external_ip": external_ip,
    }
